### ISSUE TITLE
Optimizing Clippy & linting

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust-project-goals#114]
| Metadata      | |
| --------      | --- |
| Owner(s)      | @blyxyas |
| Team(s)       | [clippy](https://github.com/rust-lang/rust-clippy) |
| Goal document | [2024h2/optimize-clippy](https://rust-lang.github.io/rust-project-goals/2024h2/optimize-clippy.html) |

## Summary

This is the formalization and documentation of the Clippy Performance Project, a project first talked about on [Zulip, July 2023](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/Clippy's.20performance). As the project consists of several points and is ever-changing, this document also has a dynamic structure and the team can add points. 

In short, this is an effort to optimize Clippy, and Rust's linting infrastructure with a point of view of making Clippy faster (both on CI/CD pipelines, and on devs' machines)

## Tasks and status

* [ ] Optimization work
  * [ ] Implementation (@blyxyas, @Alexendoo)
  * [ ] Standard reviews ([clippy](https://github.com/rust-lang/rust-clippy) ![Team][])

[Team]: https://img.shields.io/badge/Team%20ask-red

**Comment by blyxyas** [comment]
### :heavy_check_mark: New medium-sized objective completed

https://github.com/rust-lang/rust-clippy/pull/12830 *Use ControlFlow in more places*, closing https://github.com/rust-lang/rust-clippy/issues/12829 *Use ControlFlow in more visitors*

**Comment by nikomatsakis** [comment]
This issue is intended for status updates only.

For general questions or comments, please contact the owner(s) directly.

**Comment by blyxyas** [comment]
**Key developments:**

- 1 Issue open with `performance-project` label (rust-clippy#13169), we've already integrated it into the project. @Alexendoo will be taking a look at it for the time being.
- 1 performance PR merged (see above comment)
- <details><summary>We're remaking our benchmarking tool (becnhv2)</summary>Our last benchmarking tool (becnh) wasn't actually benchmarking the way we would've liked. I'm currently remaking it from scratch, this time in Rust instead of Bash. See <a href="https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/Team.20Goals/near/454795365">here</a> and <a href="https://github.com/rust-lang/rust-clippy/pull/13136#issuecomment-2249021952">here</a> for more details</details>

**What's to come:**
- After `becnhv2` is done, we'll continue working on rust#125116 (which currently needs rebasing).

**Blockers**: There are no blockers, although we need some help figuring out the main bug on rust#125116, explained [here](https://github.com/rust-lang/rust/pull/125116#issuecomment-2178589398). Someone from the compiler team taking a look at it would be **greatly** appreciated

**Comment by blyxyas** [comment]
**Key developments**:
- rust-lang/rust#125116 is now all green! :tada: We're currently doing benchmarks.
- (Clippy compilation) I'm working on moving our `declare_clippy_lint` macro from procedural to a `macro_rules` one.
- 1 Issue open (proposal) with the `performance project` label (https://github.com/rust-lang/rust-clippy/issues/13033)

**Blockers:**
There are no blockers.

**Comment by blyxyas** [comment]
**Key developments:**
    - [GnomedDev](https://github.com/GnomedDev) pushed rust-lang/rust#130553, which replaced an old Clippy infrastructure with a faster one (string matching into symbol matching)
    - Inspections into Clippy's type sizes and cache alignment are being started. Nothing fruitful yet, but I'm checking if there's something to be improved.
    - The WIP benchmarking tool for Clippy has changed the algorithm from git based to using static source code distributions (like rustup), I'm experimenting with more possibilities.

**Blockers:**
There are no blockers

**Comment by blyxyas** [comment]
# ✔️ New large objective completed

rust-lang/rust#125116 has been merged, marking half of our goal formally done!

**Comment by blyxyas** [comment]
**Key developments:**
- Discussions on https://github.com/rust-lang/rust-clippy/issues/13033#issuecomment-2501279515 are beginning to take form.
- rust-lang/rust#125116 may be contested in results. The impact may not be as large as expected, even on Clippy.
- We've been experimenting with Clippy using `rustc_driver` as a static library, instead of dynamic linking. This would be us both a way to check the performance impact of `rustc_driver` as a shared library, **and** a way to profile Clippy without filtering between `dl_*` calls.

**Blockers:** None known.
**Help wanted:** Help on [any issue marked with `performance-project`](https://github.com/rust-lang/rust-clippy/issues?q=sort%3Aupdated-desc%20is%3Aissue%20is%3Aopen%20label%3Aperformance-project) would be **greatly** appreciated!

**Comment by blyxyas** [comment]
**Key developments:**
- https://github.com/rust-lang/rust-clippy/pull/13821 is open, has been benchmarked and is in the reviewing process. This PR addresses one of the sub-goals.

**Blockers**: 
**Help wanted:** Help on [any issue marked with `performance-project`](https://github.com/rust-lang/rust-clippy/issues?q=sort%3Aupdated-desc%20is%3Aissue%20is%3Aopen%20label%3Aperformance-project) would be **greatly** appreciated! Specially https://github.com/rust-lang/rust-clippy/issues/13714

**Comment by nikomatsakis** [comment]
This is a continuing project goal, and the updates below this comment will be for the new period 2025h1

**Comment by blyxyas** [comment]
As a monthly update (previously posted on Zulip): We have some big progress!

- [rust-clippy#13821](https://github.com/rust-lang/rust-clippy/issues/13821) has been open and is currently being reviewed. It moves the MSRV logic out of lint-individualistic attribute extraction and into Clippy-wide MSRV (with a very good optimization, taking into account that only a small percentage of crates use.

- A Clippy-exclusive benchmarker has arrived, powered by the existing lintcheck infrastructure and perf. So it's compatible with flamegraph and other such tools. [rust-clippy#14194](https://github.com/rust-lang/rust-clippy/issues/14194) We can later expand this into CI or a dedicated bot.
- As you probably know, [rust#125116](https://github.com/rust-lang/rust/issues/125116) has been merged, just as a reminder of how that big goal was slayed like a dragon :dragon:.

We now know what functions to optimize (or, at least have a basic knowledge of where Clippy spends its time).
As some future functionality, I'd love to have a functionality to build cargo and rust with debug symbols and hook it up to Clippy, but that may be harder. It's not perfect, but it's a good start!

![clippy benchmark perf.data report](https://rust-lang.zulipchat.com/user_uploads/4715/1Ph9uNvqBfGDLGC55UXUhAw1/2025-02-17-000451_1918x582_scrot.png)

**Comment by blyxyas** [comment]
Monthly update!

- https://github.com/rust-lang/rust-clippy/issues/13821 has been merged. This has succesfully optimized the MSRV extraction from the source code.

On the old MSRV extraction,`Symbol::intern` use was sky high being about 3.5 times higher than the rest of the compilation combined. Now, it's at normal levels. Note that `Symbol::intern` is a very expensive and locking function, so this is very notable. Thanks to @Alexendoo for this incredible work!

As a general note on the month, I'd say that we've experimented a lot.

- Starting efforts on parallelizing the lint system.
- https://github.com/rust-lang/rust-clippy/issues/14423 Started taking a deeper look into our dependence on `libLLVM.so` and heavy relocation problems.
- I took a look into heap allocation optimization, seems that we are fine. For the moment, rust-clippy#14423 is the priority.

**Comment by blyxyas** [comment]
Monthly update!

- In [the last monthly update](https://github.com/rust-lang/rust-project-goals/issues/114#issuecomment-2730435572) we saw the impact that [interning symbols](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/symbol/struct.Symbol.html#method.intern) had on the program's performance. An effort to minimize this via a pre-interning symbol mechanism has been implemented in https://github.com/rust-lang/rust-clippy/pull/14650 and https://github.com/rust-lang/rust/pull/138682

- We're phasing out the old ["str path"](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/fn.match_def_path.html) infrastructure into a new lazy alternative. https://github.com/rust-lang/rust-clippy/pull/14705

- We're currently in the effort to optimize some documentation lints that took up to 15% of the Clippy runtime (depending on how much documentation for each line of code you had.) See https://github.com/rust-lang/rust-clippy/pull/14693

- We've also been experimenting with lots of new possibilities, mainly on parallel lints. Althought they currently are not performance improvements, there are some great hope put into them.

- Memory consumption and branch mispredictions are being monitored, they do not seem out of the ordinary.

- Monitoring cache misses and references, turns out that about 31% of cache references (792m found) are cache misses (253m found) in some benchmarks. We will check what's behind those numbers and if they can be improved.

This has been a great month for performance!
